### ISSUE TITLE
Part of email not scanned

### ISSUE THREAD

**Issue by stefaweb** [MailScanner/v5#229]
Hi!

I received this mail. A phishing mail against the OVH provider. The first link at the top is well scanned and got the red alert.

The second not. Should it or can we have a way to manage this kind of case?

![capture d ecran 2018-10-01 a 16 40 28](https://user-images.githubusercontent.com/12907102/46296070-d3ddd980-c599-11e8-9adc-7204502c58cb.png)

Config:

Debian Stretch
apache 2.4
php 7.0
mariadb 10.1
postfix 3.1.8
mailscanner 5.1.1-1
mw 1.2.10

**Comment by shawniverson** [comment]
@stefaweb Yeah, it should have flagged that.  I'll run this through once I get an instance built.

**Comment by shawniverson** [comment]
@stefaweb I am having trouble reproducing this issue.  All of my lines are getting tagged.  Is it happening to all of your emails?

```
MailScanner has detected a possible fraud attempt from "www.summitgrid.org" claiming to be www.summitgrid.com
[www.summitgrid.org]

MailScanner has detected a possible fraud attempt from "www.summitgrid.org" claiming to be www.summitgrid.net
[www.summitgrid.org]
```

**Comment by stefaweb** [comment]
@shawniverson 

I'll try to catch the complete mail next time, I'll have the same problem.

**Comment by stefaweb** [comment]
Hello @shawniverson!

I got one. Enclosed the source.
[spam.txt](https://github.com/MailScanner/v5/files/2519494/spam.txt)

**Comment by shawniverson** [comment]
@stefaweb I see some hrefs at the end but they are blank as well as the content. Are you referring to these?

**Comment by stefaweb** [comment]
@shawniverson 

I let you say that in this last example, the tag of the message worked while its structure is the same as the first example I gave. Not easy to find the problem.

**Comment by stefaweb** [comment]
Hello @shawniverson!

A new phishing email not scanned.

![capture d ecran 2018-11-07 a 18 38 42](https://user-images.githubusercontent.com/12907102/48182498-acfa8c00-e32b-11e8-8f08-3383f7ac2647.png)

[mail.source.zip](https://github.com/MailScanner/v5/files/2560596/mail.source.zip)

**Comment by stefaweb** [comment]
Another one. I received 4 of these this morning.

![sc](https://user-images.githubusercontent.com/12907102/48188017-e982b380-e33c-11e8-828a-1a895ce40ad3.png)

[mail.source.txt](https://github.com/MailScanner/v5/files/2560995/mail.source.txt)

**Comment by stefaweb** [comment]
Another one.

![capture d ecran 2018-11-09 a 11 49 06](https://user-images.githubusercontent.com/12907102/48258652-e7405800-e415-11e8-9fdf-2893a9dba0a5.png)

[mail.source.txt](https://github.com/MailScanner/v5/files/2565554/mail.source.txt)

**Comment by Skywalker-11** [comment]
Maybe a bug that breaks if there is a line break inside the link?

**Comment by stefaweb** [comment]
Yes and not only in link.

![capture d ecran 2018-11-09 a 15 08 53](https://user-images.githubusercontent.com/12907102/48267027-6db66300-e431-11e8-8052-19d6a7d49ff7.png)

**Comment by shawniverson** [comment]
@stefaweb I'm back from a conference, I'll start working on this.  I agree the line breaks are probably an issue that needs resolved in MailScanner.

**Comment by stefaweb** [comment]
Another one. Still the same problem with line breaks.

![capture d ecran 2018-11-14 a 09 51 00](https://user-images.githubusercontent.com/12907102/48470940-1595ad00-e7f3-11e8-927a-c6fb0eec7e5b.png)

![capture d ecran 2018-11-14 a 09 52 51](https://user-images.githubusercontent.com/12907102/48471000-32ca7b80-e7f3-11e8-92fb-1c3a217f87b3.png)

**Comment by shawniverson** [comment]
Multiple problems found...

1) No phishing checks for mailto: addresses (similar to the fax: and tel: issues previously found and can probably be fixed the similarly)

2) line breaks interfering with phishing checks

3) quoted-printable line continuations also interfering with phishing checks

**Comment by shawniverson** [comment]
Update....

1. Confirmed no handling of mailto: fields

2. Line breaks are being handled properly, no issue here with normal html

3. Quoted-printable is being handled properly, no issue here with line breaks

4.  Confirmed no handling of url links with plain descriptions, i.e.  "Click Here" --> www.example.org

**Comment by shawniverson** [comment]
@stefaweb @Skywalker-11 

Ok, how do we want to tackle this? This behavior is "by design" now that I have studied the code, and I don't like it.  I think it is a gaping hole in the phishing logic.

MailScanner currently deliberately ignores mailto: addresses for phishing checks and deliberately excludes them.

MailScanner also deliberately ignores any link that is not a URL.

I'm thinking about doing this...

Stop ignoring mailto: links, but provide two different behaviors depending on whether Use Stricter Phishing Net is enabled.  When this is disabled, merely verify the domain portion of the email the same way other domains are verified when this is disabled.  When this is enabled, verify the domain portion of the email is exact and also verify the username of the email is exact.

For descriptive links, I think we need a new setting, perhaps a "Highlight Hidden URLs".  When a link is found that is not a URL, display a friendly message that alerts the user to the url hidden behind the link.  Since many of these types of links are not phishing (but many are), give the user a heads up...

MailScanner has detected that "Click Here to Activate Your Account" links to "https://www.example.org" . Use caution and only open this link if it is trustworthy.

**Comment by shawniverson** [comment]
Here's a PR for mailto phishing support to try.

https://github.com/MailScanner/v5/pull/303

**Comment by shawniverson** [comment]
Here's a PR for highlighting hidden links

https://github.com/MailScanner/v5/pull/304

Translations are welcome :)

**Comment by Skywalker-11** [comment]
For mailto: maybe also include a check if the describtive text has some sort of mail address and if that is the case chexk if it matches the href part. Eg check the words before and after `@` split by space if ot matches.
rg. `<a href="mailto:something@example.com">do something for foo@example.com</a>` Should ckeck if the mail address in the href matches `foo@example.com`

**Comment by shawniverson** [comment]
@Skywalker-11 that makes sense.  Should be a trivial regex operation to incorporate this.  Should we also do something if it doesn't have an email address at all, such as a "Click here to email me?"

**Comment by msapiro** [comment]
@shawniverson The whole theory behind the 'phishing fraud' checks is to catch those things where the displayed text looks like a URL or address and doesn't match the actual domain of the target. Thus things like
```
<a href="http:/badsite.com">www.goodsite.com</a>
```
are flagged, while something like
```
<a href="http:/www.example.com/">click here</a>
```
is not. Thus, I think we should flag
```
email <a href="mailto:joe@badsite.com">joe@goodsite.com</a>
```
but not
```
<a href="mailto:joe@badsite.com">email me</a>
```
Also, mailto: links are quite different from URL links in that clicking a URL link takes you to a site whereas clicking a mailto: link just opens a mail composition window where you can see where the mail will go when you send it. That's not to say we shouldn't flag trickery like the first mailto: example; just that it's not as critical as the URLs.

**Comment by shawniverson** [comment]
@msapiro I agree that these should not be part of the Phishing Fraud logic itself, which is why I created a separate setting called "Highlight Hidden URLs."  I do see value for organizations that want to call more attention to the URLs and highlight them.  Perhaps this feature should be off by default, so that folks can opt into it if they so desire?  It is also why I changed the text to display a different message thank the phishing highlight.
